src/backend/cuda/kernel/anisotropic_diffusion.hpp

```diff
@@ -0,0 +1,239 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <backend.hpp>
+#include <common/dispatch.hpp>
+#include <Param.hpp>
+#include <debug_cuda.hpp>
+#include <math.hpp>
+#include <ops.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+static const int THREADS_X = 32;
+static const int THREADS_Y = 8;
+
+inline __device__
+int clamp(int value, int low, int high)
```

Comment:
There are a bunch of versions of clamp throughout the source code. You should try to put it in one place.

Reply:
I think it would be better do that as a separate PR, it is not related to anisotropic smoothing in particular.